### ISSUE TITLE
Is it better to backup pre-compressed data, or uncompressed?

### ISSUE THREAD

**Issue by sanmai** [restic/restic#790]
Is it better to backup pre-compressed data with restic, or uncompressed?

For example, uncompressed database dumps take many times more than compressed.

**Comment by zcalusic** [comment]
It depends a lot on your data, retention period etc...

If you compress data, it will also take less in the repository. But, if you backup that data regularly, and it changes very little deduplication could eventually do a better job in the long run.

It's best to test and see the results. For example, that's what I did when I was deciding how to backup a VM image. My first idea of using partclone first, and then backup such (smaller) image with restic was a failure. Partclone image format interfered with the deduplication somehow, so it was transferred in the whole each and every session. Eventually, I decided that backing up the raw image is more space efficient in the long run. Especially if VM image was zerofree'd first.

I still hope we will get some bad ass compression support in the restic, which could be clever about some stuff.

**Comment by sanmai** [comment]
(I answer my own question. The issue could be closed afterwards if the answer is sufficient.)

**tl;dr** No, you should avoid backing up compressed data as much as possible.

I collected some gigabytes of compressed backups, then I set up two repositories and for either of them:

    for i in $(ls backups/* | sort); do cat $i | restic backup --stdin --stdin-filename database.sql.gz; done
    for i in $(ls backups/* | sort); do zcat $i | restic backup --stdin --stdin-filename database.sql; done

Surprisingly, the repository with compressed backups took *as much space* as the backups themself:

    50G	backups/
    50G	backup-compressed/
    36G	backup-raw-uncompressed/

For the record, original backups are sized as such:

    7.1G  example.sql
    1.7G  backups/example.sql.gz

**Comment by fd0** [comment]
Thanks for posting your results, that's very interesting (although I must admit that I expected exactly this result).

To give you a bit of background: restic uses [Content Defined Chunking (CDC)](https://restic.github.io/blog/2015-09-12/restic-foundation1-cdc) to (in your use case) split the data read from stdin into variable-sized blobs. Each blob is between 512KiB and 8MiB in size, and it is identified by its (unique) SHA256 hash, so restic only saves each blob once in the repo. When the data is not compressed, restic can recognize the same blobs, decide to not save them again and reference the existing one.

When you activate compression, a small change (depending on the compression algorithm used) in the input leads to a large change in the output, so with high probability restic only finds blobs it hasn't seen before (either because the data is different, or because the cut points are in other places). Therefore it stores the new blobs alongside the existing ones and the repository grows.

So, as you already summarised above: In order to use the deduplication, try to pass uncompressed data to restic.

We have plans to implement compression for the blobs (see #21, a favorite topic for @zcalusic), but that's not yet done.

**Comment by ThomDietrich** [comment]
Hey @fd0,
I believe this is a very relevant topic and I suggest there should be an article on it in the official restic documentation.

In addition, I was wondering if you are aware of different compression algorithms (zip, bz2, gzip, xz, rar, rar5, ...) handling partial changes better than others. We have to assume that some algorithms are designed in a beneficial way in terms of blob change. Any pointers? Best!

**Comment by ThomDietrich** [comment]
I built a quick script to test this hypothesis. The script can be found here and creates multiple snapshots of a growing source directory, with different compression algorithms.

https://gist.github.com/ThomDietrich/ca4c7b943f294a4274fbc4e1d68bfb7f

The results are quite clear:

```
% du -hs *
65M	test-input
66M	test-repo-raw

172M	test-repo-bzip2
216M	test-repo-gzip
222M	test-repo-xz
```

Compression hinders restic from identifying unchanged blobs and the repo size grows way out of proportion. This is true for all three tested methods.

@fd0 to wrap this up I would be interested if restic could improve its way of finding identical blobs? It would not surprise me if the answer is No.

Best.
